### PR TITLE
LIMS XML and bug fixes

### PR DESCRIPTION
**Pull request by erikwaskiewicz**
## Change description
- Download an XML for LIMS from clicking dropdown that's now part of the download PDF button on the Report tab #60 
- Add page numbers to PDF, uses different package so env has changed, also adjusted formatting #13
- Fix issue where the wrong tab is defaulted to #47 
- Fix bug where you couldnt reopen a sample fail or enter a fail 2nd check if you're assigned to it, also tidy up of that page because there's lots of different options and there were loads of nested loops in template
- Add 'No gaps' or 'No calls' to results page when there are no reportable calls/gaps (both GUI and PDF, not needed for XML), for both DNA and RNA
- Fixed visualisation of COSMIC percentages and added columns for COSMIC counts in Depth & Report tabs & PDF (for solid tumour referrals only)
- Added 'options' page to link out to various other things such as poly lists (and probably other things in future
- Split the poly list into separate build 37 and 38 pages (you'll need to add the two poly lists on your test dev, called `build_37_polys` and `build_38_polys`) 

## Justification for minor update
No effect on processing of data, only how it's visualised

## Testing description
- Running existing unit tests
- LIMS XML not needed yet, will be tested as part of LIMS validation, may be tweaked as we go along, just review the code
- COSMIC counts/percents will be tested more in future PR - need test data from Sophie, just review the code
- Manual testing of database functionality
  - XML downloads and data matches report tab
  - PDF numbers pages correctly, data matches report tab and formatting is okay
  - No calls/ No gaps shown when no calls in report tab and PDF, instead of empty table
  - tried all possible combinations for the assigned to buttons on the samples page, shown in the comments of `view_samples.html`
  - Tested that text after hash is successfully removed when going onto the `view_samples.html` page (analyse one run normally and when you're redirected back there should be no hash)
  - Test adding polys to the build 37 and build 38 lists and make sure that they end up in the correct ones

## Test data location
N/A

## Verification results
- Unit tests all completed successfully
- Database testing listed above all performed as expected
- Deployment plan:
  - Simple model change - defaults to null
  - Notify CS about PDF changes in case there's any weird formatting that needs fixing on the fly
  - Cant specify output location for PDF/XML due to security issues, but can show CS how to set up browser to save as rather than just download straight into download folder  
  - Conda environment will need updating, test with different env name first as Viglen conda is temperamental

### CODE REVIEW

**Comment by Istaisa** [comment]
All unit tests pass.

**Comment by Istaisa** [comment]
Manual testing: XML
- **No XML download for failed samples - is this intentional?**
- Compared XML download and data for 2 passed samples - data identical between database and XML

**Comment by Istaisa** [comment]
Manual testing: PDF

- Pages number correctly
- "No calls" in variants table when no calls
- "No hotspot gaps" in gaps table when no gaps
- **For POLE gaps, all gaps show as POLE(NM_006231.3):c.NA_NA on web app and on download - this is also true of live SVD. Possibly because in the regions, formatting is e.g. POLE(NM006231.3):c.1376_p.(Ser459Phe) rather than POLE(NM006231.3):c.1376_1376_p.(Ser459Phe)**
- Data matches between download and web app - tested for 3 samples
- PDF output formatting is good

**Comment by Istaisa** [comment]
Manual Testing: Poly Lists

- **The 'tag' next to the Build 38 poly list title reads GRCh37**
- **Pages don't load without existing polylist objects - could be helpful to add this to fixtures/readme for dev**
- Same user cannot check polys - good
- **Cannot add b38 variants to b38 polylist. Tried b38 with/without chr and b37 just in case. Will investigate in code.**

File: analysis/management/commands/import.py [hunk 1]
```diff
@@ -50,20 +50,27 @@ def add_gaps_from_dict(self, gap, cutoff, new_gene_coverage_obj):
...
-                gap['percent_cosmic'] = None
-                
-        if gap['counts_cosmic'] is not None:
-            if np.isnan(gap['counts_cosmic']):
-                gap['counts_cosmic'] = None
-        
-        # error handling for COSMIC
-        if gap['percent_cosmic'] == 'N/A':
+        # handle weird inputs for COSMIC percent
+        if 'percent_cosmic' not in gaps.keys():
```
Comment (Istaisa) [inline analysis/management/commands/import.py:54]: gap not gaps

**Comment by Istaisa** [comment]
fight-209 not included in the panels_march2023 json as a b37 TSO500 DNA referral (there for b37 RNA, b38 DNA and RNA) so one sample from 37 test data does not upload

**Review by Istaisa — CHANGES_REQUESTED** [review]
Summary of changes requested:

- XML download for failed samples (EW to discuss with clinical scientists)
- POLE gaps addressed in SS pipeline update
- Minor changes to code - mostly aesthetic and 1 variable name correction, see individual comments
- Minor changes to fixtures to ensure smooth dev - polylists in fixtures, update panels with fight-209 b37 dna and aml

Otherwise all looking good!

File: analysis/templates/analysis/options_page.html [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        <h5 class="card-title">Build 37 poly list <span class="badge badge-info">GRCh37</span></h5>
+        <p class="card-text">View and edit the known poly list for <b>build 37</b>. These will be classed as polys in all assays.</p>
+        <a class="btn btn-info" href="{% url 'view_polys' 'build_37_polys' %}">View</a>
+      </div>
+    </div>
+    
+    <!-- build 38 polys -->
+    <div class="card bg-light">
+      <div class="card-body">
+        <h5 class="card-title">Build 38 poly list <span class="badge badge-success">GRCh37</span></h5>
```
[the remark this answers is not in the document]
Comment (erikwaskiewicz) [reply analysis/templates/analysis/options_page.html]: Good spot, thanks. Fixed in latest commits

File: env.yaml [hunk 3]
```diff
@@ -5,7 +5,7 @@ dependencies:
 - python=3.8
 - psycopg2=2.8
 - conda-forge::django=2.2
-- conda-forge::xhtml2pdf=0.2
+- conda-forge::python-pdfkit=1.0.
```
[the remark this answers is not in the document]
Comment (erikwaskiewicz) [reply env.yaml]: Yes, fixed in recent commits

File: analysis/management/commands/import.py (hunk 1 above)
Comment (erikwaskiewicz) [reply analysis/management/commands/import.py]: Good spot, fixed in latest commits

**Comment by erikwaskiewicz** [comment]
> XML download for failed samples (EW to discuss with clinical scientists)

- Thanks for spotting this, will discuss with CS. XML may go through a few iterations before final version is done

> POLE gaps addressed in SS pipeline update

- Yes, this is an issue with the bed2hgvs part for the current pipeline, it cant annotate POLE (and a few other genes) so just defaults to NA. New pipeline will resolve this issue

> Minor changes to code - mostly aesthetic and 1 variable name correction, see individual comments

- Done, replied to individual comments

> Minor changes to fixtures to ensure smooth dev - polylists in fixtures, update panels with fight-209 b37 dna and aml

- Polylist fixture added, it will just add the objects but no actual polys
- I've edited the poly list script and put instructions on how to fill from the poly list text files if polys are needed in the database, also includes a flag to prevent us having to double check them all for dev work
- The readme has been updated with this info and made deploy instructions clearer
- I havent updated the panels yet as I want to do a cleanup on deploy and do it all in one go, there's a few more panels in main on Viglen that aren't in here too

**Review by Istaisa — APPROVED** [review]
Tried new poly list upload - works great.
Unit tests still pass.
Merging with outstanding XML issue as this has been raised and will be developed in further releases.

**Pull request merged**